File: examples/generic_demo.py
# Generic Code Generation Example

# Generic class - Box container
class Box[T]:
    value: T = None  # Placeholder initialization

    def set(self, new_value: T):
        self.value = new_value

    def get(self) -> T:
        return self.value

# Generic function - identity
def identity[T](value: T) -> T:
    return value

# Generic function with multiple type parameters
def combine[T, U](first: T, second: U) -> str:
    return f"{first} {second}"

# Generic function with list
def get_first[T](items: list[T]) -> T:
    return items[0]

# Generic struct
struct Point[T]:
    x: T = None
    y: T = None

    def set_coords(self, x: T, y: T):
        self.x = x
        self.y = y

# Main demonstration
def main():
    # Generic classes, functions, and structs are now supported!
    # This demonstrates the code generation capability
    print("Generic code generation is working!")

File: examples/generic_demo.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using global::Sharpy.Core;

namespace Sharpy.GenericDemo
{
    public static class Exports
    {
        public class Box<T>
        {
            public T Value = null;
            public void Set(T newValue)
            {
                this.Value = newValue;
            }

            public T Get()
            {
                return this.Value;
            }
        }

        public static T Identity<T>(T value)
        {
            return value;
        }

        public static string Combine<T, U>(T first, U second)
        {
            return $"{first} {second}";
        }

        public static T GetFirst<T>(global::Sharpy.Core.List<T> items)
        {
            return items[0];
        }

        public struct Point<T>
        {
            public T X = null;
            public T Y = null;
            public void SetCoords(T x, T y)
            {
                this.X = x;
                this.Y = y;
            }
        }

        public static void Main()
        {
            global::Sharpy.Core.Exports.Print("Generic code generation is working!");
        }
    }
}
